californium-core/src/main/java/org/eclipse/californium/core/network/stack/TCPBlockwiseLayer.java

```diff
@@ -70,59 +68,84 @@ public TCPBlockwiseLayer(NetworkConfig config) {
 	 */
 	protected void handleInboundBlockwiseUpload(final Exchange exchange, final Request request) {
 
-		BlockOption block1 = request.getOptions().getBlock1();
+		if (requestExceedsMaxBodySize(request)) {
+			// allow super class to handle this error.
+			super.handleInboundBlockwiseUpload(exchange, request);
+		} else {
 
-		if ((block1 != null) && (block1.getSzx() == BERT_SZX)) {
-			KeyUri key = getKey(exchange, request);
-			Block1BlockwiseStatus status = getInboundBlock1Status(key, exchange, request);
+			BlockOption block1 = request.getOptions().getBlock1();
 
-			if (!status.addBlock(request.getPayload())) {
-				sendBlock1ErrorResponse(key, exchange, request, ResponseCode.REQUEST_ENTITY_TOO_LARGE,
-						"body exceeded expected size " + status.getBufferSize());
-			}
+			if ((block1 != null) && (block1.getSzx() == BERT_SZX)) {
+				KeyUri key = getKey(exchange, request);
+				Block1BlockwiseStatus status = getInboundBlock1Status(key, exchange, request);
+				
+				if (block1.getNum() == 0 && status.getCurrentNum() > 0) {
+					status = resetInboundBlock1Status(key, exchange, request);
+				}
 
-			status.setCurrentNum(status.getCurrentNum() + (request.getPayloadSize() / BERT_INT_BLOCK_SIZE));
-			if (block1.isM()) {
+				if (block1.getNum() != status.getCurrentNum()) {
+					// ERROR, wrong number, Incomplete
+					LOGGER.log(Level.WARNING,
+							"peer sent wrong block, expected no. {0} but got {1}. Responding with 4.08 (Request Entity Incomplete)",
+							new Object[] { status.getCurrentNum(), block1.getNum() });
 
-				// Message has more blocks.
+					sendBlock1ErrorResponse(key, exchange, request, ResponseCode.REQUEST_ENTITY_INCOMPLETE,
+							"wrong block number");
 
-				LOGGER.log(Level.FINE, "acknowledging incoming block1 [num={0}], expecting more blocks to come",
-						block1.getNum());
+				} else if (!status.hasContentFormat(request.getOptions().getContentFormat())) {
 
-				Response piggybacked = Response.createResponse(request, ResponseCode.CONTINUE);
-				piggybacked.getOptions().setBlock1(block1.getSzx(), true, block1.getNum());
-				piggybacked.setLast(false);
+					sendBlock1ErrorResponse(key, exchange, request, ResponseCode.REQUEST_ENTITY_INCOMPLETE,
+							"unexpected Content-Format");
 
-				exchange.setCurrentResponse(piggybacked);
-				lower().sendResponse(exchange, piggybacked);
+				} else if (!status.addBlock(request.getPayload())) {
+					sendBlock1ErrorResponse(key, exchange, request, ResponseCode.REQUEST_ENTITY_TOO_LARGE,
+							"body exceeded expected size " + status.getBufferSize());
+				}
 
-			} else {
```

Comment:
Though this commit is based on an other commit, its hard to see the resulting change.
Therefore, depending on the amount and complexity of the changes, it's good to rebase an "squash" commits (squash means to create a commit, which collects the resulting changes of several commits. Depending on the git tools your using, there are several ways to squash).

Removing `} else {` may be wrong, the stuff below may then be executed, even if above an error was detected. 
But I'm unsure about that, because this is a change of a change and it's not obvious what is the  resulting change.